### ISSUE TITLE
postgis geometry

### ISSUE THREAD

**Issue by jfgodoy** [balderdashy/sails-postgresql#43]
I need to use a geometry field of postgis. I know that waterline can't have a so specific type.. but what about a work around?

```
attributes: {
  x: 'float',
  y: 'float',  
  geom: {
    type: 'string',
    geometry: {
      def: 'SRID=4326;Point',
      bind: ['x', 'y'],
      saveBindData: true
    }
  }
}
```
The idea is, if an attribute has a geometry property, then the adapter will ignore the type and will build the column definition using geometry.def. If saveBindData is true then the adapter will build the column definitions for that fields too, in other case its information will be used to build the geom data, but not will be saved in the database.

The bind variable can be also an array of coordinates following the geojson spec according to the type defined in geometry.def, or a string using the WKT spec, or even a json field using geojson spec.

what do you think about this approach?

**Comment by tjwebb** [comment]
@jfgodoy where is your fork?

**Comment by jfgodoy** [comment]
It's in:
https://github.com/jfgodoy/sails-postgresql/tree/geometrySupport
https://github.com/jfgodoy/waterline-schema/tree/geometrySupport
https://github.com/jfgodoy/waterline-sequel/tree/geometrySupport

It's not up to date and doesn't have test :(.

**Comment by tjwebb** [comment]
Do you have the time to write some tests? I'm seeing more interest in this feature, and would happily merge a PR from you if you can clean it up a bit. Let me know, thanks.

**Comment by kevbaker** [comment]
So it isn't possible to use Waterline  "Custom Types" to make a geometry data-type then?

**Comment by tjwebb** [comment]
> @jfgodoy, do we just need to rebase your geometrySupport feature and create some tests for it?

Probably. This would provide initial support for geom types.

In the long term I'd love to support deeper PostGIS features as well, probably in an extension.

**Comment by jfgodoy** [comment]
I just do the rebase on my forks. I do not have the time to do tests.
Theses are the changes I did to get a basic support for geometry types. A lot of work is required to make it robust, so please, only use my changes as reference.

https://github.com/jfgodoy/waterline-sequel/tree/geometrySupport
- modify select builder to get geometry columns as text
- replace `returning *` on queries to allow return geometry columns as text
- add geometry search criteria. I only added intersects

https://github.com/jfgodoy/sails-postgresql/tree/geometrySupport
- allow type `geometry(<geomType>, <srid>)`. This need to be improved, I only check the word geometry.

https://github.com/jfgodoy/waterline/tree/geometrySupport
- allow type `geometry(<geomType>, <srid>)`. This need to be improved, I only check the word geometry.
- add validation for geometry values. I did not.

Also geometry columns require an spatial index to improve query performance.
The implementation only support EWKT format, wich is a PostGIS specific format.

Examples of use:
```
/* api/models/Point.js */
module.exports = {
  connection: 'somePostgresqlServer',
  attributes: {
    id: 'integer',
    geom: 'geometry(Point, 4326)'
  }
};
```
Now you can insert geometry data:
```
POST http://localhost:1337/api/point
{
"id":1,
"geom": "SRID=4326;Point(70 33)"
}
```
And retrieve all the point inside a given area
```
GET http://localhost:1337/point?where={"geom":{"intersects": "SRID=4326;Polygon((-50 -10, -50 -50 ,-75 -50, -75 -10, -50 -10))"  }}
```

**Comment by kevbaker** [comment]
This is great Jorge.

I'll start by implementing your changes so I understand them better. Then
should have some time to help with the tests.

**Comment by bwestergard** [comment]
Any movement on this?

**Comment by kevbaker** [comment]
I've been using the native method on models for now. I actually switched to
Mongodb for a few reasons. Not as fast as PostGIS, but does offer many of
the key geospatial query features I needed as of version 2.6.





On Thu, Jul 30, 2015 at 12:58 PM, Bjørn Westergard <notifications@github.com

**Comment by dbkaplun** [comment]
+1

**Comment by warpedgeoid** [comment]
+4326

**Comment by sailsbot** [comment]
Thanks for posting, @jfgodoy.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
Reopening until a PR to the roadmap is made.  @tjwebb if I remember correctly, didn't you make some headway on this?

**Comment by sailsbot** [comment]
Thanks for posting, @jfgodoy.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by parallaxisjones** [comment]
Any Updates on this?

**Comment by tjwebb** [comment]
The waterline-postgresql adapter has PostGIS support: https://github.com/waterlinejs/postgresql-adapter

**Comment by parallaxisjones** [comment]
@tjwebb thanyou for the reply! :)
